### PR TITLE
ui/termstatus: Quote funny filenames

### LINKED ISSUE #2260: restic should sanitize filenames in output
Output of `restic version`
--------------------------
restic 0.9.5 compiled with go1.12.4 on linux/amd64

How did you run restic exactly?
-------------------------------
```
$ truncate -s 1M $'x\e[31m\nred\nis\nbeautiful\nand\nnewlines\nare\nannoying\n'
$ restic -r repo init
$ restic -r repo backup x*
$ restic -r repo ls latest
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
not relevant

Expected behavior
-----------------
Output not messing up the terminal.

Actual behavior
---------------
I got the following output in the terminal with parts of it being red due to the escape sequence included in the filename:
```
$ restic -r repo backup x*                                                     
repository 5d1e95b3 opened successfully, password is correct
[0:00] 1 files 1024.000 KiB, total 1 files 1024.000 KiB, 0 errors
/x
red
is
beautiful
and
newlines

[0:00] 1 files 1024.000 KiB, total 1 files 1024.000 KiB, 0 errors
/x
red
is
beautiful
and
newlines
Files:           1 new,     0 changed,     0 unmodified
[0:00] 1 files 1024.000 KiB, total 1 files 1024.000 KiB, 0 errors
/x
red
is
beautiful
and
newlines
Dirs:            0 new,     0 changed,     0 unmodified
Added to the repo: 512.500 KiB

processed 1 files, 1024.000 KiB in 0:00
snapshot 0e0b6872 saved
$ restic -r repo ls latest
repository 5d1e95b3 opened successfully, password is correct
snapshot 0e0b6872 of [/tmp/tmp.XgxPfBNlHs/x
red
is
beautiful
and
newlines
are
annoying
] filtered by [] at 2019-04-30 23:45:43.594055711 +0200 CEST):
/x
red
is
beautiful
and
newlines
are
annoying


```

Steps to reproduce the behavior
-------------------------------
See "How did you run restic exactly?"

Do you have any idea what may have caused this?
-----------------------------------------------
restic outputs filenames without any kind of sanitization so any included escape sequences are interpreted by the terminal and special characters like `\n` can mess up the output.

Do you have an idea how to solve the issue?
-------------------------------------------
Sanitize filenames when outputting them to a terminal.

Did restic help you or made you happy in any way?
-------------------------------------------------
Yes, it's a pleasure to use, especially when compared to some enterprise™ backup "solutions".

### LINKED ISSUE #4191: Carriage returns in file names cause corrupted verbose output
Output of `restic version`
--------------------------
restic 0.15.0 compiled with go1.19.5 on darwin/arm64

How did you run restic exactly?
-------------------------------

```
export RESTIC_PASSWORD_COMMAND='security find-generic-password -a (redacted) -s (redacted) -w';
export GOOGLE_PROJECT_ID='(redacted)';
export GOOGLE_APPLICATION_CREDENTIALS="(redacted)";
export RESTIC_REPOSITORY='gs:(redacted):/';

caffeinate -i restic -v -v backup "$HOME"
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
GCP Coldline storage

Expected behavior
-----------------
If a filename contains a carriage return, it should be sanitized before printing to the terminal.

Actual behavior
---------------

When backing up file names containing carriage returns, the verbose output gets corrupted because the carriage returns are output directly to the terminal, sending the following output to the beginning of the lines. Here's a a real example. Dunno why, but I have a folder whos name contains a carriage return character: `Linux Admin Security Guide<CR>`. The verbose output prints like:

```
/servers/, saved in 0.041s (0 B added, 0 B stored, 0 B metadata)in Security Guide
/software/, saved in 0.002s (0 B added, 0 B stored, 0 B metadata)n Security Guide
/system/, saved in 0.002s (0 B added, 0 B stored, 0 B metadata)min Security Guide
/users/, saved in 0.002s (0 B added, 0 B stored, 0 B metadata)dmin Security Guide
/viruses/, saved in 0.002s (0 B added, 0 B stored, 0 B metadata)in Security Guide
/vpn/, saved in 0.001s (0 B added, 0 B stored, 0 B metadata) Admin Security Guide
/, saved in 0.120s (0 B added, 0 B stored, 0 B metadata)inux Admin Security Guide
modified  /Users/q/Bibliotheque/technology/Unix/_books/Linux Admin survival guide/, saved in 0.020s (0 B added, 0 B stored, 0 B metadata)
modified  /Users/q/Bibliotheque/technology/Unix/_books/Linux From Scratch/, saved in 0.118s (0 B added, 0 B stored, 0 B metadata)
```

These first seven lines should appear more like:

```
modified  /Users/q/Bibliotheque/technology/Unix/_books/Linux Admin Security Guide/servers/, saved in 0.1s (0 B added, 0 B stored, 0 B metadata)
…
```

Steps to reproduce the behavior
-------------------------------

1. Add a carriage return to a filename
2. Run `restic backup` with verbose output enabled.

Do you have any idea what may have caused this?
-----------------------------------------------

Special characters in file names are printed as-is to the terminal.

Do you have an idea how to solve the issue?
-------------------------------------------

Special characters in file names should be sanitized before printing to the terminal.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Restic is awesome. Thanks!

### PR DESCRIPTION
**Pull request by greatroar**
Fixes #2260, #4191.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by MichaelEischer** [comment]
Would that also solve #2260?

**Comment by greatroar** [comment]
@MichaelEischer Updated to fix that issue too.

File: internal/ui/backup/text.go [hunk 1]
```diff
@@ -144,3 +148,13 @@ func (b *TextProgress) Finish(snapshotID restic.ID, start time.Time, summary *Su
 		ui.FormatDuration(time.Since(start)),
 	)
 }
+
+// Quote filenames with funny characters in them.
+func quote(filename string) string {
+	for _, r := range filename {
+		if r == ' ' || r == unicode.ReplacementChar || !unicode.IsPrint(r) {
```
Comment (MichaelEischer) [inline internal/ui/backup/text.go:155]: What is the idea behind also quoting spaces and the unicode replacement characters? Both won't mess up the terminal. That would also make the output change much less invasive.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Could you add a very short changelog entry?

File: internal/ui/backup/text.go (hunk 1 above)
Comment (greatroar) [reply internal/ui/backup/text.go:155]: I can drop the quoting of spaces, but the Unicode replacement char is what you get when the input is not UTF-8.

File: internal/ui/backup/text_test.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		{" foo ", true},
+		{"foo bar", true},
+		{"foo\nbar", true},
+		{"foo\rbar", true},
+		{"foo\abar", true},
+		{"\xff", true},
+		// Since this is for human consumption, we don't quote-escape '\\'.
+		{`c:\foo\bar`, false},
+		// Issue #2260.
+		{"\x31m red is beautiful", true},
```
[the remark this answers is not in the document]
Comment (greatroar) [reply internal/ui/backup/text_test.go:26]: Fixed, thanks.

File: internal/ui/backup/text.go [hunk 3]
```diff
@@ -62,7 +64,7 @@ func (b *TextProgress) Update(total, processed Counter, errors uint, currentFile
 
 	lines := make([]string, 0, len(currentFiles)+1)
 	for filename := range currentFiles {
-		lines = append(lines, filename)
+		lines = append(lines, quote(filename))
```
[the remark this answers is not in the document]
Comment (greatroar) [reply internal/ui/backup/text.go]: Yes. Moved quoting there.

File: internal/ui/backup/text.go (hunk 1 above)
Comment (greatroar) [reply internal/ui/backup/text.go]: Dropped quoting of spaces, i.e., of the ASCII space. Tabs and other whitespace are still quoted.

**Comment by greatroar** [comment]
Added the changelog entry.

**Comment by MichaelEischer** [comment]
I'm sorry to have another question, but isn't the change to `internal/ui/backup/text.go` quote the `item` passed to `func (b *TextProgress) CompleteItem(messageType, item string, previous, current *restic.Node, s archiver.ItemStats, d time.Duration)` still necessary to solve #4191? `Terminal.SetStatus` only covers the status bar, but not the other terminal output.

**Comment by MichaelEischer** [comment]
@greatroar  Ping?

**Comment by greatroar** [comment]
Yes, you're right. Fixed now. Sorry for the delay.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**